services/bots/github_apps.py

```diff
@@ -0,0 +1,242 @@
+        and app.is_configured()
+        and (
+            # If there is a repo we want only the apps that cover said repo
+            (repository and app.is_repo_covered_by_integration(repository))
+            # If there is no repo we still need some true value
+            or (not repository)
+        )
+    )
+
+
+def _get_apps_from_weighted_selection(
+    owner: Owner, installation_name: str, repository: Optional[Repository]
+) -> List[GithubAppInstallation]:
+    """This function returns an ordered list of GithubAppInstallations that can be used to communicate with GitHub
+    in behalf of the owner. The list is ordered in such a way that the 1st element is the app to be used in Torngit,
+    and the subsequent apps are selected as fallbacks.
+
+    IF the repository is provided, the selected apps also cover the repo.
+    IF installation_name is not the default one, than the default codecov installation
+      is also selected as a possible fallback app.
+
+    Apps are selected randomly but assigned weights based on how recently they were created.
+    This means that older apps are selected more frequently as the main app than newer ones.
+    (up to 10 days, when the probability of being chosen is the same)
+    The random selection is done so we can distribute request load more evenly among apps.
+    """
+    # Map GithubAppInstallation.id --> GithubAppInstallation
+    ghapp_installations_filter: Dict[int, GithubAppInstallation] = {
+        obj.id: obj
+        for obj in filter(
+            lambda obj: _can_use_this_app(obj, installation_name, repository),
+            owner.github_app_installations or [],
+        )
+    }
+    # We assign weights to the apps based on how long ago they were created.
+    # The idea is that there's a greater chance that a change misconfigured the app,
+    # So apps recently created are selected less frequently than older apps
+    keys = list(ghapp_installations_filter.keys())
+    weights = [
+        min(
+            MAX_GITHUB_APP_SELECTION_WEIGHT,
+            _get_installation_weight(ghapp_installations_filter[key]),
+        )
+        for key in keys
+    ]
+    # We pick apps one by one until all apps have been selected
+    # Obviously apps with a higher weight have a higher change of being selected as the main app (1st selection)
+    # But it's important that others are also selected so we can use them as fallbacks
+    apps_to_consider = []
+    apps_to_select = len(keys)
+    selections = 0
+    while selections < apps_to_select:
+        selected_app_id = random.choices(keys, weights, k=1)[0]
+        apps_to_consider.append(ghapp_installations_filter[selected_app_id])
+        # random.choices chooses with replacement
+        # which we are trying to avoid here. So we remove the key selected and its weight from the population.
+        key_idx = keys.index(selected_app_id)
+        keys.pop(key_idx)
+        weights.pop(key_idx)
+        selections += 1
```

Comment:
It doesn't seem to have weights...
OR if I use the "weight" as the "count", is that statistically equivalent?...

Reply:
(I commented before from the wrong browser :E apologies)

OK so looking at this again it would not work.
Because it's missing the weights, and using "weight" as counts would re-introduce the problem of replacement.
We want _all_ available apps to be selected with the weights taken into consideration.